async_asgi_testclient/tests/asgi_spec/test_http.py

```diff
@@ -0,0 +1,391 @@
+"""ASGI spec http tests
+
+Tests to verify conformance with the ASGI specification. This module tests
+the http protocol.
+
+These tests attempt to make sure that TestClient conforms to
+the ASGI specification documented at
+https://asgi.readthedocs.io/en/latest/specs/main.html
+"""
+from async_asgi_testclient import TestClient
+from multidict import CIMultiDict
+from urllib.parse import quote
+
+import pytest
+
+
+@pytest.mark.asyncio
+async def test_asgi_version_is_present_in_http_scope(mock_app):
+    """
+    The key scope["asgi"] will also be present as a dictionary containing a scope["asgi"]["version"] key
+    that corresponds to the ASGI version the server implements.
+    https://asgi.readthedocs.io/en/latest/specs/main.html#applications
+
+    We expect this to be version 3.0, as that is the current spec version.
+    """
+
+    async def handle_all(scope, receive, send):
+        if scope["type"] == "http":
+            assert "asgi" in scope
+            assert scope["asgi"]["version"] == "3.0"
+
+    mock_app.handle_all = handle_all
+
+    async with TestClient(mock_app) as client:
+        await client.get("/")
+
+
+@pytest.mark.asyncio
+async def test_http_spec_version_is_missing_or_correct(mock_app):
+    """
+    asgi["spec_version"] (Unicode string) – Version of the ASGI HTTP spec this server
+     understands; one of "2.0" or "2.1". Optional; if missing assume 2.0
+    https://asgi.readthedocs.io/en/latest/specs/www.html#http-connection-scope
+
+    TestClient doesn't specify spec_version at the moment, which is also okay.
+    Note that if newer features are added (specifically allowing None in 'server' scope value)
+    then the spec_version needs to be passed correctly.
+    """
+
+    async def handle_all(scope, receive, send):
+        if scope["type"] == "http":
+            assert (
+                "spec_version" not in scope["asgi"]
+                or scope["asgi"]["spec_version"] == "2.0"
+            )
+
+    mock_app.handle_all = handle_all
+
+    async with TestClient(mock_app) as client:
+        resp = await client.get("/")
+        assert resp.status_code == 200
+
+
+@pytest.mark.asyncio
+async def test_http_version_is_1_1(mock_app):
+    """
+    http_version (Unicode string) – One of "1.0", "1.1" or "2".
+    https://asgi.readthedocs.io/en/latest/specs/www.html#http-connection-scope
+    TestClient currently only supports http/1.1, so test that it is set correctly.
+    """
+
+    async def handle_all(scope, receive, send):
+        if scope["type"] == "http":
+            assert scope["http_version"] == "1.1"
+
+    mock_app.handle_all = handle_all
+
+    async with TestClient(mock_app) as client:
+        resp = await client.get("/")
+        assert resp.status_code == 200
+
+
+@pytest.mark.xfail(
+    AssertionError, reason="TestClient does not uppercase scope['method']"
+)
```

Comment:
Good catch! I think we should fix it. Did you try to fix it and see if the integration test with some frameworks still pass?

Reply:
No, I didn't touch this one - I thought this was up to you as maintainer to decide when is the right time if it will changed, because it will potentially break existing usage, so you might want to schedule it for an appropriate minor version with a breaking change notice or whatever makes sense for the package.

Reply:
I'm planning to release your changes in v2.0.0 and document these breaking changes in the CHANGELOG. 

I guess all the major framworks already uppercase the HTTP method, and that's why this is working even though it's not correct with the specification.